packages/api/src/payments/paymentProvider.ts

```diff
@@ -129,32 +131,94 @@ export abstract class BasePaymentProvider implements PaymentProvider {
+    }
+
+    // update payment provider
     if (intentState.customerID && payment.invoiceID) {
-      const invoice = await loaders.invoicesByID.load(payment.invoiceID)
-      if (!invoice) throw new Error(`Invoice with ID ${payment.invoiceID} does not exist`)
-
-      const subscription = await dbAdapter.subscription.getSubscriptionByID(invoice.subscriptionID)
-      if (!subscription)
-        throw new Error(`Subscription with ID ${invoice.subscriptionID} does not exist`)
-
-      const user = await dbAdapter.user.getUserByID(subscription.userID)
-      if (!user) throw new Error(`User with ID ${subscription.userID} does not exist`)
-
-      // adding or updating paymentProvider customer ID for user
-      const paymentProviderCustomers = user.paymentProviderCustomers.filter(
-        ppc => ppc.paymentProviderID !== this.id
-      )
-      paymentProviderCustomers.push({
-        paymentProviderID: this.id,
-        customerID: intentState.customerID
-      })
-      await dbAdapter.user.updatePaymentProviderCustomers({
-        userID: user.id,
-        paymentProviderCustomers
-      })
+      await this.updatePaymentProvider(dbAdapter, subscription, intentState.customerID)
     }
     return updatedPayment
   }
+
+  /**
+   * adding or updating paymentProvider customer ID for user
+   * @param dbAdapter
+   * @param subscription
+   * @param customerID
+   * @private
+   */
+  private async updatePaymentProvider(
+    dbAdapter: DBAdapter,
+    subscription: OptionalSubscription,
+    customerID: string
+  ) {
+    if (!subscription) {
+      throw new Error('Empty subscription within updatePaymentProvider method.')
+    }
+    const user = await dbAdapter.user.getUserByID(subscription.userID)
+    if (!user) throw new Error(`User with ID ${subscription.userID} does not exist`)
+
+    const paymentProviderCustomers = user.paymentProviderCustomers.filter(
+      ppc => ppc.paymentProviderID !== this.id
+    )
+    paymentProviderCustomers.push({
+      paymentProviderID: this.id,
+      customerID
+    })
+    await dbAdapter.user.updatePaymentProviderCustomers({
+      userID: user.id,
+      paymentProviderCustomers
+    })
+  }
+
+  /**
+   * Create regular user out of temp user and activate subscription
+   * @param dbAdapter
+   * @param userID
+   * @param subscription
+   * @private
+   */
+  private async activateTempUserAndSubscription(
+    dbAdapter: DBAdapter,
+    userID: string,
+    subscription: OptionalSubscription
+  ): Promise<OptionalSubscription> {
+    // create non-temp user id
+    userID = removePrefixTempUser(userID)
+    const tempUser = await dbAdapter.tempUser.getTempUserByID(userID)
+    if (!tempUser) {
+      throw new Error('User not found while updating payment with intent state.')
+    }
+
+    // creating the new user out of the existing temp_user
+    const user = await dbAdapter.user.createUserFromTempUser(tempUser)
```

Comment:
Is it possible that we run this subscription activation for 2 subscriptions with the same temp user and will try to create user twice? And actually succeed as the email uniqueness validation is only in the mutation.

Reply:
Not possible. The mutation only allows to create one subscription. If an user would try to register an new subscription it creates a new temp user. And it's not possible to add a subscription for a temp user via the editor. We should be fine ;)